### ISSUE TITLE
Update Library functionality

### ISSUE THREAD

**Issue by RRohrssen** [unfoldingWord-dev/ts-desktop#451]
May require UI changes...

**Comment by RRohrssen** [comment]
Waiting for TS API v3.0

**Comment by RRohrssen** [comment]
Blocked by TS API v3.0

**Comment by RRohrssen** [comment]
Chris did a good summary of how this works in Slack. Here it is for posterity...

Chris Whitfield [12:24 AM] 
@sharon: @r_rohrssen Clarification: Future versions of ts, like current versions, will always come populated with the current list of languages from tD.  At some point, tS-Desktop, like Android has now, will come with an "Update Library" option, where a user can initiate a download of a more up-to-date list, if changes to tD have been made.  An active connection is only needed during the download of the updates.  After that, the app will contain the updates internally and can be used fully offline.  This new feature will definitely NOT be in version 8 yet, but may be ready by version 9.

**Comment by cdwhitfield73** [comment]
The language list update part of this has been completed in a separate issue.  The rest of this is blocked by waiting for the new api.

**Comment by cckozie** [comment]
Koz will review the UI on Android with Christine and make recommendations for possible changes vs implementing it as is.

**Comment by neutrinog** [comment]
@cckozie will review the android implementation and discuss with @ChrisJarka about potential UI changes. Then @neutrinog will write up the new specification.

**Comment by cdwhitfield73** [comment]
Blocked until #819 is complete

**Comment by cckozie** [comment]
We need better error handling when processing the updates. 
If there is no Internet connection the Update Target Languages and Update Chunk Markers both indicate that they completed successfully. If we don't have an Internet connection we should alert the user to that situation before attempting the update.
If the Internet connection is broken during the update, all four options indicate that they completed successfully.

**Comment by cckozie** [comment]
Current UI:
![image](https://cloud.githubusercontent.com/assets/8171759/19164714/be1f0740-8bc6-11e6-8bf2-bf93fe8ff4fb.png)
Run times (minutes:seconds) on 3 WIndows test machines (Mac is faster):
Update Target Languages - 0:30 to 1:08
Update Sources - 0:44 to 1:47
Update Chunk Markers - 2:12 to 5:55
Update All - 3:20 to 7:22

**Comment by cckozie** [comment]
![targets1](https://cloud.githubusercontent.com/assets/8171759/19166270/8519c42a-8bcc-11e6-8f9f-d40f09889ef5.png)
![targets2](https://cloud.githubusercontent.com/assets/8171759/19166273/88fe8256-8bcc-11e6-881b-7c08e701848a.png)
![sources 1](https://cloud.githubusercontent.com/assets/8171759/19166283/8d3b9c50-8bcc-11e6-8b78-9b7b2bf3f473.png)
![sources 2](https://cloud.githubusercontent.com/assets/8171759/19166293/94508abe-8bcc-11e6-9d90-ea68a667e298.png)
![chunks 1](https://cloud.githubusercontent.com/assets/8171759/19166303/9e2e0322-8bcc-11e6-9ac2-c39a7175ec78.png)
![chunks 2](https://cloud.githubusercontent.com/assets/8171759/19166307/a255b9b8-8bcc-11e6-907c-e812272fb963.png)
![all](https://cloud.githubusercontent.com/assets/8171759/19166316/a68416e2-8bcc-11e6-9d42-b1df57069ff5.png)

**Comment by cdwhitfield73** [comment]
The error handling is in place in the app.  This is an issue with the door43-client.  It was catching the "no connection" errors internally and passing back a "positive" result to the app.  Joel is going to fix this in his module.

**Comment by ChrisJarka** [comment]
Regarding "Updating Languages" and "updating sources" - 
Does "languages" mean **target languages**? Is there space in the message to add clarification?  
Does "sources" mean **source language biblical text**? What about the tHelps- are they part of the sources that are updated? If not then is there space in the message to add clarification?  If so, then the verbiage is good for sources.

**Comment by ChrisJarka** [comment]
If I understand the process correctly, this download is just the database of resources and not the data. If so then I have just a thought - since updating the sources takes some time - what about adding additional verbiage (since they have time to read the message :)) something like: You may need to do additional downloading of specific resources. Check the list of resources in your projects for further information.

Something like this message might help those that go do a download and then go back to no connectivity to do the work. It might be inconvenient/frustrating for them to then realize that the updates are not really there.

**Comment by ChrisJarka** [comment]
As for "chunk markers" - why cannot the process of correcting chunk markers run at the same time the import of .txt or .usfm files are imported? Why are we asking the end-user to do two processes to have usable data.
Just asking. :)

**Comment by ChrisJarka** [comment]
Just a really picky thing - can the zero have a slash through it? So it doesn't look so much like an o. 
Is the zero-slash internationally understood?

**Comment by cckozie** [comment]
See new suggested text in comments above with the screen shots. Remove the options to update the chunks and update all. (suggested text still needs approval)

**Comment by cckozie** [comment]
@ChrisJarka - Please review and finalize the text desired on these dialogs.
![capture1](https://cloud.githubusercontent.com/assets/8171759/19566051/94968c8e-96ae-11e6-9487-18c537ccd065.PNG)
![capture2](https://cloud.githubusercontent.com/assets/8171759/19566060/9c002016-96ae-11e6-85ae-5a63e2809512.PNG)
![capture3](https://cloud.githubusercontent.com/assets/8171759/19566071/a5bbc7c2-96ae-11e6-99e8-110d6e70cc19.PNG)
![capture4](https://cloud.githubusercontent.com/assets/8171759/19566087/acfb2f50-96ae-11e6-993c-2764fef3ca92.PNG)
![capture5](https://cloud.githubusercontent.com/assets/8171759/19566098/b3907d02-96ae-11e6-9446-54589594b9f3.PNG)

**Comment by ChrisJarka** [comment]
For the 'Update Options' menu card
I think you might mean 'this requires an Internet connection' instead of 'this will use your internet connection'.

For the Download List of New and Update Source Texts - Thank you for using my suggested dialog. 
This only updates the list of available source texts. After the download, an Internet connection will be required to update the content of the source text used in a Project.
This process may take some time.

BUT - now that I read it separated I would like to suggest a change or two - so it would read:
This will only update the list of available source texts. This process may take some time.
After the download, an Internet connection will be required to update the content of the source text used in a Project.

**Comment by ChrisJarka** [comment]
For the "Updating Source Texts" message:
Maybe something like the suggested change above?

Updating list of source texts. Please remember, this process will only update the list of Source Texts. It may take some time.

**Comment by ChrisJarka** [comment]
I like the word 'no' instead of the zero digit when there have been no updates. Much better!

**Comment by cdwhitfield73** [comment]
All the above modifications have been made.

**Comment by cckozie** [comment]
When creating a new project and performing a download during the source language selection, there are a couple problems.
1. Any source languages selected before the download get unselected
2. It would be nice if the dialog displayed during the download would appear above the source language selection dialog rather than closing and then reopening the source selection dialog.

**Comment by cdwhitfield73** [comment]
I can definitely fix number 1.  As for number 2, I realized when I built it that the flow was not ideal, but necessary for two reasons.  The format we are using does not allow having two modals open at once, so I must close the source modal to open the loading modal.  Also, after the download, reloading the source modal is what allows the buttons to update based on the new status.  If this is a "deal breaker", let me know.

**Comment by cdwhitfield73** [comment]
Not sure why this is still blocked, but all the issues above have been fixed.

**Comment by cckozie** [comment]
Ran update list of available source texts this morning. It indicated 81 new source texts available. I don't think that are all appearing in the list of what is available to download, and some that do appear also have problems.
![image](https://cloud.githubusercontent.com/assets/8171759/21611185/5d3bb674-d191-11e6-975b-e01ae51e4d85.png)

**Comment by cckozie** [comment]
I see a similar result after reinstalling the app. (This was all done on OSX)
Also, there are 26 languages listed in the Download Available Source Texts option, but there are 41 rows in the source_language table of index.sqlite.

**Comment by cckozie** [comment]
It looks like this has been a problem since build 93. It did not exist in build 87.

**Comment by cdwhitfield73** [comment]
The issue with blank lines showing up on some languages has been fixed.  This was a code issue, not a data issue.  As for the languages in the database not showing up in the list, this is because they have no level 3 checked resources to display.
